src/Router/Artsy2.tsx

```diff
@@ -0,0 +1,48 @@
+import { ContextProps as _ContextProps } from "Components/Artsy"
+import React, { SFC } from "react"
+import { getUser } from "Utils/getUser"
+import { createEnvironment } from "../Relay/createEnvironment"
+
+// FIXME: Move from Artsy to Artsy2
+export interface ContextProps extends _ContextProps {
+  // FIXME: Type properly
+  mediator?: any
+  system?: any
+}
+
+const Context = React.createContext<ContextProps>({})
+
+/**
+ * Creates a new Context.Provider with a user and Relay environment, or defaults
+ * if not passed in as props.
+ */
+export const ContextProvider: SFC<ContextProps> = ({ children, ...props }) => {
+  const currentUser = getUser(props.currentUser)
+  const relayEnvironment =
+    props.relayEnvironment || createEnvironment({ user: currentUser })
+
+  const providerValues = {
+    ...props,
+    currentUser,
+    relayEnvironment,
+  }
+
+  return <Context.Provider value={providerValues}>{children}</Context.Provider>
+}
+
+export const ContextConsumer = Context.Consumer
+
+/**
+ * A HOC utility function for injecting renderProps into a component.
+ */
+export const withContext = Component => {
+  return props => {
+    return (
+      <ContextConsumer>
+        {contextValues => {
+          return <Component {...contextValues} {...props} />
+        }}
+      </ContextConsumer>
+    )
+  }
+}
```

Comment:
I think the types for this context need to be adjusted such that some of them are optional for users of the provider but not for users of the consumer. E.g. a relay env can optionally be passed to the provider, but a consumer will always have one.

Reply:
How should I express this?

Reply:
Looking at the context typings you’ve shown above:

<img width="403" alt="screen shot 2018-08-23 at 12 56 40 pm" src="https://user-images.githubusercontent.com/236943/44548813-03b0dc00-a6d4-11e8-8941-77ab471f3380.png">

It seems like you’ll need to separately use the `Provider` and the `Consumer` types. E.g.

```tsx
const Context = React.createContext({})

export const Provider: React.Provider<{ mediator: Mediator, relayEnvironment?: Relay.Environment }> = Context.Provider

export const Consumer: React.Consumer<{ mediator: Mediator, relayEnvironment: Relay.Environment }> = Context.Consumer
```

----

Seeing as repeating the list could get cumbersome if they are identical except for some of them being optional in the provider interface, you could do something like the following to only make a subset optional:

```ts
type X = "foo" | "bar"

interface Foo {
    foo: string
    bar: string
    baz: string
}

type PartialSubset<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>> & Partial<Pick<T, K>>

type Y = PartialSubset<Foo, "foo" | "bar">

const y: Y = { baz: "baz" }
```